autoload/ale/engine.vim

```diff
@@ -0,0 +1,218 @@
+" Stores information for each job including:
+"
+" linter: The linter dictionary for the job.
+" buffer: The buffer number for the job.
+" output: The array of lines for the output of the job.
+let s:job_info_map = {}
+
+function! ale#engine#GetJobID(job) abort
```

Comment:
Most of these functions should probably be script functions with `s:`, as they aren't intended to be visible anywhere else, unless there's some kind of performance reason to name them otherwise.